### ISSUE TITLE
Question: Log multiple Table Storage rows for each EntityFrameworkEvent.Entries

### ISSUE THREAD

**Issue by egaspar** [thepirat000/Audit.NET#195]
Hi,

I'm using the EF extension alongside the Azure Table Storage provider, and it's working fine for single operations.

In the EntityMapper call, when configuring the table storage provider, is there a way to create a separate log entry for each EF Entry that was affected by the current database operation?

```
Audit.Core.Configuration.Setup()
.UseAzureTableStorage(_ => _
.ConnectionString(connectionString)
.TableName(e => "foo")
.EntityMapper(e =>
{
  // I want to create a log for each entry
  foreach (var x in ((AuditEventEntityFramework)e).EntityFrameworkEvent.Entries)
  {
  }
}
```

**Comment by thepirat000** [comment]
The EF extension generates one audit event per `SaveChanges` call, and only if there were changes. So you will not be able to save more than one row just by using the `EntityMapper` on the `AzureTableStorage`.

But you should be able to create your custom DataProvider, furthermore you can inherit from the `AzureTableDataProvider` and just override what is needed (in your case the `InsertEvent`/`InsertEventAsync` methods.

For example:

```
public class MyCustomAzureTableProvider : AzureTableDataProvider
{
    public override object InsertEvent(AuditEvent auditEvent)
    {
        var efEvent = auditEvent.GetEntityFrameworkEvent();
        var entries = efEvent.Entries;
        object result = null;
        foreach (var entry in entries)
        {
            // override audit event's Entries list to have only this entry
            efEvent.Entries = new List<EventEntry>() { entry };
            // insert the event
            result = base.InsertEvent(auditEvent);
        }
        return result;
    }
    
    public async override Task<object> InsertEventAsync(AuditEvent auditEvent)
    {
        // similar to InsertEvent but calling base.InsertEventAsync
    }
}
```

Then you can configure the provider as follows (note the `EntityMapper` function will receive an EF audit event with just one entry on it):

```
Audit.Core.Configuration.Setup()
    .UseCustomProvider(new MyCustomAzureTableProvider(az => az
        .ConnectionString("cnnstring")
        .TableName("table")
        .EntityMapper(e => new AuditEventTableEntity(e))));
```

Note you should use the latest version released some minutes ago (14.0.3) that one contains the constructor accepting the fluent configuration API.

**Comment by egaspar** [comment]
Worked like a charm, thanks for your help!
